server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesIndexAction.java

```diff
@@ -117,16 +151,160 @@ protected FieldCapabilitiesIndexResponse shardOperation(final FieldCapabilitiesI
                 }
             }
         }
-        return new FieldCapabilitiesIndexResponse(shardId.getIndexName(), responseMap);
+        return new FieldCapabilitiesIndexResponse(request.index(), responseMap, true);
     }
 
-    @Override
-    protected Writeable.Reader<FieldCapabilitiesIndexResponse> getResponseReader() {
-        return FieldCapabilitiesIndexResponse::new;
+    private boolean canMatchShard(FieldCapabilitiesIndexRequest req) throws IOException {
+        if (req.indexFilter() == null || req.indexFilter() instanceof MatchAllQueryBuilder) {
+            return true;
+        }
+        assert req.nowInMillis() != 0L;
+        ShardSearchRequest searchRequest = new ShardSearchRequest(req.shardId(), req.nowInMillis(), AliasFilter.EMPTY);
+        searchRequest.source(new SearchSourceBuilder().query(req.indexFilter()));
+        return searchService.canMatch(searchRequest).canMatch();
     }
 
-    @Override
-    protected ClusterBlockException checkRequestBlock(ClusterState state, InternalRequest request) {
-        return state.blocks().indexBlockedException(ClusterBlockLevel.METADATA_READ, request.concreteIndex());
+    private ClusterBlockException checkGlobalBlock(ClusterState state) {
+        return state.blocks().globalBlockedException(ClusterBlockLevel.READ);
+    }
+
+    private ClusterBlockException checkRequestBlock(ClusterState state, String concreteIndex) {
+        return state.blocks().indexBlockedException(ClusterBlockLevel.READ, concreteIndex);
+    }
+
+    /**
+     * An action that executes on each shard until it finds a replica that can match the provided
```

Comment:
Saying "replica" confused me because it made me think that we might check multiple copies of the same shard in some cases, while the below logic suggests that we just check one copy for every shard until we find one that matches. Should this comment say "shard" instead of "replica" in this comment?

Reply:
++, I pushed https://github.com/elastic/elasticsearch/pull/57276/commits/fb3007c0832b7a5eb604e76978fb0047d70ecd61